tc/autotuner/options_cache-inl.h

```diff
@@ -0,0 +1,342 @@
+  typename Backend::OptionsCacheProtoType buf;
+  struct stat buffer = {0};
+  if (stat(filename.c_str(), &buffer) == 0) {
+    std::ifstream serialized(filename, std::ios::binary);
+    buf.ParseFromIstream(&serialized);
+  }
+  fromProtobuf(buf);
+}
+
+template <typename Backend>
+void OptionsCache<Backend>::storeCacheToFile(
+    const std::string& filename) const {
+  // toProtobuf() takes the lock too, get a copy of the result first
+  auto proto = toProtobuf();
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    std::fstream serialized(
+        filename, std::ios::binary | std::ios::trunc | std::ios::out);
+    if (!serialized.is_open()) {
+      LOG(ERROR) << "Failed to open the output stream for dumping protobuf: "
+                 << filename;
+    } else {
+      proto.SerializePartialToOstream(&serialized);
+    }
+  }
+}
+
+template <typename Backend>
+void OptionsCache<Backend>::recordRuntime(
+    const lang::CanonicalTcString& tc,
+    const std::vector<TensorInfo>& inputs,
+    const std::vector<TensorInfo>& outputs,
+    const std::string& deviceStr,
+    const typename Backend::MappingOptionsType& options,
+    Duration duration) {
+  std::lock_guard<std::mutex> lock(mutex);
+  ++numberCacheAttempts;
+  OptionsCacheKey key{tc, inputs, outputs, deviceStr};
+  bool inserted = false;
+  auto range = store_.equal_range(key);
+  for (auto it = range.first; it != range.second; ++it) {
+    if (it->second.mappingOptions == options) {
+      it->second.runtimes.push_back(duration);
+      inserted = true;
+      break;
+    }
+  }
+  if (!inserted) {
+    store_.emplace(
+        key,
+        OptionsCacheValue<Backend>{std::vector<Duration>{duration}, options});
+  }
+}
+
+namespace detail {
+template <typename Backend>
+struct OptionsWithMedianAndRuntimes {
+  Duration median;
+  std::vector<Duration> runtimes;
+  typename Backend::MappingOptionsType mappingOptions;
+};
+
+template <typename Backend>
+std::vector<OptionsWithMedianAndRuntimes<Backend>> sortedOptions(
+    const OptionsCacheKey& key,
+    const std::unordered_multimap<
+        OptionsCacheKey,
+        OptionsCacheValue<Backend>,
+        OptionsCacheKeyHash>& store) {
+  using Options = OptionsWithMedianAndRuntimes<Backend>;
+  std::vector<Options> toSort;
+  if (store.count(key) == 0) {
+    return {};
+  }
+  auto range = store.equal_range(key);
+  for (auto it = range.first; it != range.second; ++it) {
+    if (it->second.runtimes.size() == 0) {
+      throw std::runtime_error("No runtime for cache option");
+    }
+    toSort.push_back(Options{median(it->second.runtimes),
+                             it->second.runtimes,
+                             it->second.mappingOptions});
+  }
+  std::sort(
+      toSort.begin(), toSort.end(), [](const Options& a, const Options& b) {
+        // fun with C++, a.median < b.median does not mix with templates
+        return operator<(a.median, b.median);
```

Comment:
Is this something compiler specific?
`a.median < b.median` seems to work fine on my system.

Reply:
My guess is that the way `std::sort` is implement in some version of the standard library prevents ADL from working properly.

Reply:
it fails with 5.4.1